### PR TITLE
stack-lock translator & haskell builder

### PR DESCRIPTION
**Pull request by DavHau**
Implements a builder for haskell and a translator for stack.yaml.lock files.

This is a rough POC which needs to be improved in many places.

### CODE REVIEW

File: src/subsystems/haskell/translators/stack-lock/utils.nix [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        echo "all-cabal-hashes" seems to be outdated
+        exit 1
+      fi
+    '';
+  in
+    pkgs.runCommand "cabal-json-files" {}
+    (l.concatStringsSep "\n"
+      (l.map (c: convertOne c.name c.version) candidates));
+
+  batchCabalData = candidates: let
```
Comment (m1-s) [inline src/subsystems/haskell/translators/stack-lock/utils.nix:67]: we could add a doc string to each of the functions in this file

File: src/subsystems/haskell/builders/default/default.nix [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+          doBenchmark = false;
+
+          libraryToolDepends = libraryHaskellDepends;
+          executableHaskellDepends = libraryHaskellDepends;
+          testHaskellDepends = libraryHaskellDepends;
+          testToolDepends = libraryHaskellDepends;
+
+          libraryHaskellDepends =
+            (with pkgs.haskell.packages.ghc8107; [
+              # Cabal_3_2_1_0
```
Comment (m1-s) [inline src/subsystems/haskell/builders/default/default.nix:89]: whats the meaning of this comment?

File: src/subsystems/haskell/translators/stack-lock/default.nix [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+      projectSource = "${tree.fullPath}/${project.relPath}";
+      projectTree = tree.getNodeFromPath project.relPath;
+
+      # parse the cabal file
+      cabalFiles =
+        l.filter
+        (l.hasSuffix ".cabal")
+        (l.attrNames projectTree.files);
+
+      cabalFile = projectTree.getNodeFromPath (l.head cabalFiles);
```
[the remark this answers is not in the document]
Comment (DavHau) [reply src/subsystems/haskell/translators/stack-lock/default.nix:116]: If there is no cabal file, the translator cannot work. Failing in this case is correct. Anyways, the discoverer will not even call this translator if the cabal file is missing.

**Comment by DavHau** [comment]
I will merge this for now. I think maintenance should continue on main branch. This seems usable enough. I'm happy to continue to incorporate any feedback from this thread.

**Pull request merged**